python/mim_control/dg/wbc.py

```diff
@@ -0,0 +1,144 @@
+        for endeff_name in self.endeff_names:
+            self.dg_robot.createPosition('pos_' + endeff_name, endeff_name)
+
+        # Create the relative leg positions.
+        # TODO: Handle feet offset here as well.
+        self.com_signal = self.dg_robot.signal('com')
+
+        self.rel_eff_pos = subtract_vec_vec(
+            hom2pos(self.dg_robot.signal('pos_' + self.endeff_names[0])),
+            self.com_signal)
+
+        for i in range(1, len(self.endeff_names)):
+            self.rel_eff_pos = stack_two_vectors(
+                self.rel_eff_pos,
+                subtract_vec_vec(
+                    hom2pos(self.dg_robot.signal('pos_' + self.endeff_names[i])),
+                    self.com_signal
+                ),
+                3 * i, 3
+            )
+
+        dg.plug(add_vec_vec(
+                self.wcom_pd_ctrl.wrench_sout,
+                self.w_com_ff_sout
+            ), self.f_ctrl.w_com_sin)
+        dg.plug(self.rel_eff_pos, self.f_ctrl.relative_position_endeff_sin)
+
+
+        ###
+        # Impedance controllers.
+        self.imps = []
+
+        # Create plugable vector signals to control the impedance controller.
+        for i, endeff_name in enumerate(self.endeff_names):
+            imp = mim_control_dg.ImpedanceController(prefix + '_imp_' + endeff_name)
+            imp.initialize(pin_robot.model, 'universe', endeff_name)
+
+            dg.plug(
+                stack_two_vectors(
+                    selec_vector(self.f_ctrl.forces_sout, 3*i, 3*(i + 1)),
+                    zero_vec(3, ''),
+                    3, 3
+                ),
+                imp.feed_forward_force_sin)
+
+            self.imps.append(imp)
+
+        # The final computed control.
+        self.joint_torques_sout = self.imps[0].joint_torque_sout;
+        for i in range(1, self.ne):
+            self.joint_torques_sout = add_vec_vec(
+                    self.joint_torques_sout, self.imps[i].joint_torque_sout)
+
+        ###
+        # Export all the signals for the user of the PyEntity.
+        self.kc_sin = self.wcom_pd_ctrl.kp_com_sin
+        self.dc_sin = self.wcom_pd_ctrl.kd_com_sin
+        self.kb_sin = self.wcom_pd_ctrl.kp_base_sin
+        self.db_sin = self.wcom_pd_ctrl.kd_base_sin
+
+        self.des_com_pos_sin = self.wcom_pd_ctrl.desired_com_position_sin
+        self.des_com_vel_sin = self.wcom_pd_ctrl.desired_com_velocity_sin
+        self.des_ori_pos_sin = self.wcom_pd_ctrl.desired_base_orientation_sin
+        self.des_ori_vel_sin = self.wcom_pd_ctrl.desired_base_angular_velocity_sin
+
+        self.cnt_array_sin = self.f_ctrl.cnt_array_sin
+
+
+    def plug(self, robot, base_position, base_velocity):
+        # Args
+        #   robot; DGM robot device
+        #   base_position: The base position as a 7 dim vector signal
+        #   base_velocity: The base velocity as a 6 dim vector signal
+
+        # Create the input to the dg_robot.
+        base_pose_rpy = basePoseQuat2PoseRPY(base_position)
```

Comment:
Where is this imported from?

Reply:
Ok from dg_tools, it is not intuitive with the `import *`

Reply:
Thanks for pointing this out. I made the imports from `dg_tools` explicit.